Add support for GNOME 43

Moon-0xff opened the pull request:
It simply adds the [removed code](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/gnome-42/js/ui/panel.js#L332) for `AggregateLayout` to `widgets.js`

This is a quick and simple fix for [my own testing purposes](https://github.com/Moon-0xff/gnome-mpris-label/issues/7), so don't expect much of it :P

Moon-0xff committed:
Add support for GNOME 43

JasonLG1979 commented:
Sorry I missed this. Must have got lost in the noise. I will test this out after work today and if all goes well merge it. Thanks.

ChrisLauinger77 commented:
![Bildschirmfoto vom 2022-12-06 18-05-17](https://user-images.githubusercontent.com/3518684/205976407-39b1c0aa-dbe3-41ee-ad47-9cd9acfceab2.png)
thats how it looks like for me

ChrisLauinger77 commented:
its called quicksettings
https://gjs.guide/extensions/upgrading/gnome-shell-43.html#quick-settings

ChrisLauinger77 commented:
I just found they have QuickSettingsLayout as replacement.
https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/quickSettings.js
I think I can make another proposal how to solve this tomrrow

Moon-0xff commented:
Using QuickSettingsLayout it looks like this:
![Screenshot from 2022-12-06 17-04-08](https://user-images.githubusercontent.com/66807763/206011878-2ea8c9bd-d0bc-4b07-bf49-1181e81edb67.png)

I just copied the entire thing to `widgets.js`, from [line 381](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/quickSettings.js#L381) to 579, added GLib, and replaced `AggregateLayout` to `QuickSettingsLayout`

This isn't a clean way of doing this but probably maintains compatibility with older versions of GNOME.

JasonLG1979 commented:
That doesn't look right either though.

Moon-0xff commented:
That's the commit of this PR

JasonLG1979 commented:
I'm pretty sure it depends on a sty class to set the width.

Moon-0xff commented:
`AggregateLayout` gets the width from `vfunc_get_preferred_width`. You can set the height creating a function named `vfunc_get_preferred_height` too.

Moon-0xff commented:
Both functions are [present in QuickSettings](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/quickSettings.js#L512), however they are a lot more complicated.

JasonLG1979 commented:
The layout doesn't set the wisth it just makes sure everything stays the proper size. The Aggregate menu style class sets the width. I'm looking for a replacement for that if it exists.

JasonLG1979 commented:
If a suitable replacement does not exist we can create our own. I'd not do that though. I'd rather it be a built-in style class so the extension matched  whatever theme more.

quaintlogic commented:
I might be totally missing something here, messing around locally I noticed that commenting out line #1427 [(on PR #69) ](https://github.com/JasonLG1979/gnome-shell-extension-mpris-indicator-button/blob/44981472a998b1f7ece0bdf39c341cbd0d185bc2/mprisindicatorbutton%40JasonLG1979.github.io/widgets.js#L1427) fixes the width issue.

This part could be totally essential and I'm missing something, tried Rhythmbox, YT Music and Youtube - all seem to work well.

doronbehar commented:
I tested this patch along with commenting out line 1427 as @giinuu suggested and it's working great for me. Thanks for everyone who tested and contributing :) Would be great for others if this got a bit of attention.

petar-v commented:
Hi! Thanks for that extension, it's great! I have missed having it! :) 

I just tested it on my end as well. Looks alright IMHO, and it does the job.
![image](https://user-images.githubusercontent.com/3801306/221482716-e9a86413-9231-4715-870c-1e12b69ad383.png)

Moreover, this is tested on Gnome v44. It works perfectly as long as the `metadata.json` includes that version. :)

Moon-0xff commented:
Commenting 1427 negates this PR, indeed is the only change the extension needs to run on 43/44

Mershl commented:
> Commenting 1427 negates this PR, indeed is the only change the extension needs to run on 43/44

Is it still working for you on 44 @Moon-0xff ? Fedora 38 throws an JS error for me:

```
JS ERROR: TypeError: Gtk.IconTheme.get_default is not a function
```

Location: https://github.com/JasonLG1979/gnome-shell-extension-mpris-indicator-button/blob/master/mprisindicatorbutton%40JasonLG1979.github.io/dbus.js#L1915

~Someone know an alternative to `Gtk.IconTheme.get_default()` or is it just gone with Gtk4?~
-> https://gjs.guide/extensions/upgrading/gnome-shell-44.html#gtk-icontheme

Moon-0xff commented:
@Mershl  I can't test right now but probably you are commenting the wrong line. If you are starting from JasonLG1979:master the line number is 1428. The line to comment is this one: `this.menu.box.set_layout_manager(new AggregateLayout());`

Mershl commented:
> @Mershl I can't test right now but probably you are commenting the wrong line. If you are starting from JasonLG1979:master the line number is 1428. The line to comment is this one: `this.menu.box.set_layout_manager(new AggregateLayout());`

I'm seeing a seperate issue on 44 regarding the default IconTheme, that is also mentioned in the porting guide of 44 (linked in my previous post). I was just curious why it was working for you on 44, but maybe this interface change was introduced late.

Moon-0xff commented:
I tested this when 44 was still in beta so yeah, might be a change introduced after my tests. Have you tried to replace the `Gtk.IconTheme` statements to `St.IconTheme`?

Mershl commented:
```
        const iconTheme = new St.IconTheme();
        return this.app_id && (iconTheme ? iconTheme.has_icon(iconName) : false) ? Gio.ThemedIcon.new(iconName) : null;
```
seems to work nicely on 44. I'm not sure when St12, the first version of this API supporting `.IconTheme.has_icon` was released though. This could likely break for 43 and lower.
